File: easy/remove-duplicates-from-sorted-array.ts
// The first submission -> Wrong
/*
function removeDuplicates(nums: number[]): number {
    let len = nums.length;
    let p = 1;
    let temp;
    while (p < len) {
        if (nums[p] > nums[p - 1]) {
            p++;
        } else if (nums[p] < nums[p - 1]) {
            temp = nums[p];
            nums[p] = nums[p - 1];
            nums[p - 1] = temp;
        } else {
            if (p + 1 < len) {
                temp = nums[p];
                nums[p] = nums[p + 1];
                nums[p + 1] = temp;
            }
            len--;
        }
    }
    return len;
};
*/

// The second submission -> 193ms, 48.7mb
function removeDuplicates(nums: number[]): number {
    const used = {};
    let usedCount = 0;
    
    const len = nums.length;
    let p = 0;
    while (p < len) {
        if (used[nums[p]]) {
            nums[p] = Infinity;
        } else {
            used[nums[p]] = true;
            usedCount++;
        }
        p++;
    }
    
    nums.sort((a, b) => a - b);
    
    return usedCount;
};

File: easy/remove-duplicates-from-sorted-array.java
// The fourth submission -> 12ms, 48.7mb
class Solution {
    public int removeDuplicates(int[] nums) {
        Map<Integer, Boolean> used = new HashMap<Integer, Boolean>();
        int usedCount = 0;

        int len = nums.length;
        int p = 0;
        while (p < len) {
            if (used.containsKey(nums[p])) {
                nums[p] = Integer.MAX_VALUE;
            } else {
                used.put(nums[p], true);
                usedCount++;
            }
            p++;
        }

        for (int i = 0; i < usedCount; i++) {
            for (int j = i; j < len; j++) {
                if (nums[j] != Integer.MAX_VALUE) {
                    if (nums[i] != nums[j]) {
                        nums[i] = nums[j];
                        nums[j] = Integer.MAX_VALUE;
                    }
                    break;
                }
            }
        }

        return usedCount;
    }
}